From 533ca4c0e73e768cd82fe76fb30f814043f1f718 Mon Sep 17 00:00:00 2001
Message-ID: <533ca4c0e73e768cd82fe76fb30f814043f1f718.1765628489.git.lorenzo@kernel.org>
From: Lorenzo Bianconi <lorenzo@kernel.org>
Date: Sat, 13 Dec 2025 12:29:16 +0100
Subject: [nf-next 01/11] net: airoha: Move net_devs registration in a
 dedicated routine

Since airoha_probe() is not executed under rtnl lock, there is small race
where a given device is configured by user-space while the remaining ones
are not completely loaded from the dts yet. This condition will allow a
hw device misconfiguration since there are some conditions (e.g. GDM2 check
in airoha_dev_init()) that require all device are properly loaded from the
device tree. Fix the issue moving net_devices registration at the end of
the airoha_probe routine.

Fixes: 9cd451d414f6e ("net: airoha: Add loopback support for GDM2")
Signed-off-by: Lorenzo Bianconi <lorenzo@kernel.org>
---
 drivers/net/ethernet/airoha/airoha_eth.c | 39 ++++++++++++++++--------
 1 file changed, 26 insertions(+), 13 deletions(-)

--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -3084,19 +3084,29 @@ static int airoha_alloc_gdm_port(struct
 	if (airhoa_is_phy_external(port)) {
 		err = airoha_setup_phylink(dev);
 		if (err)
-			goto free_metadata_dst;
+			return err;
 	}
 #endif
+	return 0;
+}
 
-	err = register_netdev(dev);
-	if (err)
-		goto free_metadata_dst;
+static int airoha_register_gdm_devices(struct airoha_eth *eth)
+{
+	int i;
 
-	return 0;
+	for (i = 0; i < ARRAY_SIZE(eth->ports); i++) {
+		struct airoha_gdm_port *port = eth->ports[i];
+		int err;
+
+		if (!port)
+			continue;
 
-free_metadata_dst:
-	airoha_metadata_dst_free(port);
-	return err;
+		err = register_netdev(port->dev);
+		if (err)
+			return err;
+	}
+
+	return 0;
 }
 
 static int airoha_probe(struct platform_device *pdev)
@@ -3187,6 +3197,10 @@ static int airoha_probe(struct platform_
 		}
 	}
 
+	err = airoha_register_gdm_devices(eth);
+	if (err)
+		goto error_napi_stop;
+
 	return 0;
 
 error_napi_stop:
@@ -3200,16 +3214,18 @@ error_hw_cleanup:
 	for (i = 0; i < ARRAY_SIZE(eth->ports); i++) {
 		struct airoha_gdm_port *port = eth->ports[i];
 
-		if (port && port->dev->reg_state == NETREG_REGISTERED) {
-			unregister_netdev(port->dev);
-			airoha_metadata_dst_free(port);
+		if (!port)
+			continue;
+
+		airoha_metadata_dst_free(port);
 #if defined(CONFIG_PCS_AIROHA)
-			if (airhoa_is_phy_external(port)) {
-				phylink_destroy(port->phylink);
-				airoha_pcs_destroy(port->pcs);
-			}
-#endif
+		if (port->phylink) {
+			phylink_destroy(port->phylink);
+			airoha_pcs_destroy(port->pcs);
 		}
+#endif
+		if (port->dev->reg_state == NETREG_REGISTERED)
+			unregister_netdev(port->dev);
 	}
 	free_netdev(eth->napi_dev);
 	platform_set_drvdata(pdev, NULL);
